dg.go

```diff
@@ -244,14 +244,16 @@ func (d *directedGraph[NodeType]) HasReadyNodes() bool {
 }
 
 func (d *directedGraph[NodeType]) PopReadyNodes() map[string]ResolutionStatus {
+	result := make(map[string]ResolutionStatus)
 	d.lock.Lock()
+	defer d.lock.Unlock()
 	// Transfer the map to a local variable to minimize time locked, and reset the graph's value.
 	readyMap := d.readyForProcessing
 	d.readyForProcessing = make(map[string]*node[NodeType])
-	d.lock.Unlock()
```

Comment:
Now that we're holding the lock for the whole function execution, transferring the map doesn't "minimize time locked", so you should at least fix the comment.

However, since we're now holding the lock throughout, I bet it's more efficient to clear `readyForProcessing` when we're done than to delete entries as we go or to create a new map.  Something like this,
```go
	result := make(map[string]ResolutionStatus)
	d.lock.Lock()
	defer d.lock.Unlock()
	for _, node := range d.readyForProcessing {
		result[node.ID()] = node.status
	}
	clear(d.readyForProcessing)
	return result
```

Reply:
Done.